### PR TITLE
[REVIEW] Serialize an empty column table with non zero rows [skip ci]

### PR DESCRIPTION
**Pull request by kuhushukla**
`JCudfSerialization` should work in cases we have a table with no columns but non zero number of rows - which is a valid use case when using `count` operator.

### CODE REVIEW

File: java/src/main/java/ai/rapids/cudf/JCudfSerialization.java [hunk 1]
```diff
@@ -720,6 +720,9 @@ private static SerializedTableHeader calcHeader(ColumnBufferProvider[] columns,
                                                   int numRows) {
     DType[] types = new DType[columns.length];
     long[] nullCount = new long[columns.length];
+    if (columns.length == 0) {
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/JCudfSerialization.java]: This should be before the arrays are heap-allocated, or alternatively this check shouldn't be here and instead `getSlicedSerializedDataSizeInBytes` is updated to handle an empty columns array.

File: java/src/main/java/ai/rapids/cudf/JCudfSerialization.java [hunk 2]
```diff
@@ -1276,11 +1279,16 @@ private static Table sliceUpColumnVectors(SerializedTableHeader header,
...
       int numColumns = dataTypes.length;
-      ColumnVector[] vectors = new ColumnVector[numColumns];
+      ColumnVector[] vectors = new ColumnVector[Math.max(1,numColumns)];
       DeviceMemoryBuffer validity = null;
       DeviceMemoryBuffer data = null;
       DeviceMemoryBuffer offsets = null;
       try {
+        if (numColumns == 0) {
+          vectors[0] = new ColumnVector(DType.INT32, numRows, Optional.of(0L),
+              DeviceMemoryBuffer.allocate(DType.INT32.sizeInBytes * numRows), validity, offsets);
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/JCudfSerialization.java]: Do we really want to allocate `numRows` of INT32 values when they're not going to be used?  Ideally we shouldn't be allocating anything, including a libcudf `table`, in this scenario.  If we _really_ have to allocate something then it should be a `BOOL8` column.

Comment (kuhushukla) [reply java/src/main/java/ai/rapids/cudf/JCudfSerialization.java]: I tried out passing nulls for buffers and non zero rows and it does not return the right values. I will again look into this as I agree allocation is totally synthetic and was not my first choice.

**Comment by revans2** [comment]
There was a discussion earlier about cudf supporting empty tables at the C++ layer, and they decided that they didn't want to.  I would feel inclined to follow suit and actually switch this around and add in an assertion at

https://github.com/rapidsai/cudf/blob/c190442ca216085cec47913de9bb7a08f3b01d2b/java/src/main/java/ai/rapids/cudf/Table.java#L55-L56

and any place in JCudfSerialization that takes ColumnVectors directly instead of a Table.

If we try to support empty tables it will end up being a pain to special case it everywhere in the java API when we cannot count on the native layer to support it.  I would suggest that we try to work around it at a higher level if possible.  If not we can come back and make this all work if needed.

**Comment by kuhushukla** [comment]
@jlowe , @revans2 , thank you for the comments, let me go with the suggested approach and update the PR asap.

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by kuhushukla** [comment]
@jlowe , @bobby does this version align with your previous comments? Appreciate any inputs.

**Comment by kuhushukla** [comment]
> Latest changes lgtm, but I'd like to see a unit test that exercises the new ability to encode and decode only a row count with no columns.

Added with a minor fix to flush call.

**Comment by kuhushukla** [comment]
@jlowe Added tests. Request for review. Thanks a lot!

**Comment by kuhushukla** [comment]
Updating changelog (again) here in a bit.  @revans2 , @razajafri  could u help take a second look at the change in the mean time?

**Comment by jlowe** [comment]
@kkraus14 this is ready to be admin merged.  It's not tested by CI.  Thanks!

**Comment by kuhushukla** [comment]
Thanks much @jlowe , @kkraus14 !

**Pull request merged**